torch/_inductor/codegen/cpp_micro_gemm.py

```diff
@@ -531,44 +528,41 @@ class CppMicroGemmAMX(CppMicroGemm):
     alignas(4096) {{input_t}} dequantized_B_buf[buf_size];
     {%- endif %}
 
-    const auto b_tile_ptr_stride = ldb * {{vnni_size}};
-
-    auto load_B_row = [&]({{input2_t}}* {{restrict_keyword}} src, {{input_t}}* {{restrict_keyword}} dst) {
-        auto b_int8 = at::vec::Vectorized<int8_t>::loadu(src, static_cast<int64_t>(32));
-        auto b_bf16 = at::vec::convert<{{input_t}}>(b_int8);
-        b_bf16.store(dst);
-    };
-
-    auto load_B_tile = [&]({{input2_t}}* B_ptr, int idx, int num_b_rows) {
-        {{input_t}}* base_addr = dequantized_B_buf + idx;
-        {{kernel.unroll_pragma(8)}}
-        for (int i = 0; i < num_b_rows; i++) {
-            load_B_row(
-                B_ptr + i * b_tile_ptr_stride,
-                base_addr + i * 32
-            );
-        }
-    };
-    auto load_dequantized_B = [&](int n) {
+    auto load_dequantized_B = [&](int base_idx) {
         // Load a tile of B & cache it in L1D.
-        {{kernel.unroll_pragma(4)}}
-        for (int k = 0; k < K; k += {{block_k}}) {
-            int num_b_rows = (k < last_k_offset) ? 16 : tail_k_size;
-            {{kernel.unroll_pragma(2)}}
-            for (int tile_col = 0; tile_col <= 1; tile_col++) {
-                load_B_tile(
-                    const_cast<{{input2_t}}*>(B) + n + k * ldb + tile_col * {{16 * vnni_size}},
-                    (k / {{block_k // 2}} + tile_col) * num_elements_per_b_tile,
-                    num_b_rows
-                );
-            }
+        // The assumption here is that if N would be a multiple of block_n.
+        {{input2_t}}* base_addr = const_cast<{{input2_t}}*>(B) + base_idx;
+        for (int idx_dq = 0, idx_q = 0; idx_dq < buf_size; idx_q += ldb, idx_dq += {{block_n}}) {
+        {%- if block_n in [16, 32] %}
+            auto b_int8 = at::vec::Vectorized<int8_t>::loadu(
+                base_addr + idx_q,
+                static_cast<int64_t>({{block_n}})
+            );
+            auto b_bf16 = at::vec::convert<{{input_t}}>(b_int8);
+            b_bf16.store(dequantized_B_buf + idx_dq);
+        {%- elif block_n == 48 %}
```

Comment:
Hi @jgong5, please advise if there's any way to add error checking to ensure `block_n` is either 16, 32 or 48. Thanks!

Reply:
Maybe check `block_n` here to decide using `use_cached_dequantized_B` or not

https://github.com/pytorch/pytorch/blob/8b55f649610eab61ad3504bd0685da97dd95809e/torch/_inductor/codegen/cpp_micro_gemm.py#L728

Reply:
Maybe check the `block_n` in `codegen_define` when decide to use `use_cached_dequantized_B`.

Reply:
This seems over-complicated and also makes the logic limited to handle 16, 32 and 48. Can you refer to the following code (in `CppMicroGemmFP32Vec`) to simplify and generalize your implementation? Basically, there are two parts, i.e. 32-element chunks and the tail (<32).
```c++
    for (int64_t m = 0; m < M; m += {{block_m}}) {
        int64_t block_m = std::min<int64_t>(M - m, {{block_m}});
        for (int64_t n = 0; n < N; n += {{block_n}}) {
            if (block_m == {{block_m}}) {
                {{kernel_name}}_kernel<{{block_m}}, {{block_n}}, accum>(
                    A + m * lda,
                    B + n,
                    C + m * ldc + n,
                    K,
                    lda,
                    ldb,
                    ldc
                );
            } else {
                switch (block_m) {
{%- for b in range(block_m - 1, 0, -1) %}
                case {{b}}:
                    {{kernel_name}}_kernel<{{b}}, {{block_n}}, accum>(
                        A + m * lda,
                        B + n,
                        C + m * ldc + n,
                        K,
                        lda,
                        ldb,
                        ldc
                    );
                    break;
{%- endfor %}
                default:
                    {{kernel.assert_function}}(false, "Unsupported block_m: {{block_m}}");
                }
            }
        }
    }
```

Reply:
Done, thanks!